cms/core/blocks/markup.py

```diff
@@ -105,3 +108,31 @@ def render(self, value: dict, context: dict | None = None) -> Union[str, "SafeSt
         """The Wagtail core TableBlock has a very custom `render` method. We don't want that."""
         rendered: str | SafeString = super(blocks.FieldBlock, self).render(value, context)
         return rendered
+
+
+class ONSTableBlock(TinyTableBlock):
+    """The ONS table block."""
+
+    source = blocks.CharBlock(label=_("Source"), required=False)
+    footnotes = blocks.RichTextBlock(label=_("Footnotes"), features=settings.RICH_TEXT_BASIC, required=False)
+
+    def get_context(self, value: dict, parent_context: dict | None = None) -> dict:
+        """Insert the DS-ready options in the template context."""
+        context: dict = super().get_context(value, parent_context=parent_context)
+
+        data = value.get("data", {})
+        return {
+            "title": value.get("title"),
+            "options": {
+                "caption": value.get("caption"),
+                "ths": data.get("headers", []),
+                "trs": [{"tds": row} for row in data.get("rows", [])],
+            },
+            "source": value.get("source"),
+            "footnotes": value.get("footnotes"),
+            **context,
+        }
```

Comment:
While I appreciate the idea here, we tend to have the values passed to the macro directly in the template files. It has the advantage that you can see what are the used options at a glance when looking at the template files. 
Since `options` don't really have a justification until it gets to the macro - let's return a flat dict here and construct the dict in the template files.

Reply:
I second Mebin, but could be biased from the main menu work. But I do side with avoiding logic in templates as much as possible. I like Dan's approach!